File: backend/app/schemas/extracted_content.py
"""Paginated extracted containers, blocks, tables, and search responses."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import Field

from app.models.extracted_block import ExtractedBlockType
from app.models.extracted_container import ExtractedContainerType
from app.models.language_block_result import LanguageCode
from app.schemas.base import ApiSchema
from app.schemas.common import PaginationData

ExtractedContentSource = Literal["NATIVE", "OCR"]


class ExtractedContainerResponse(ApiSchema):
    """Lightweight metadata for paginated container API responses."""

    id: UUID
    extraction_run_id: UUID
    container_type: ExtractedContainerType
    container_index: int = Field(ge=0)
    name: str | None = None
    title: str | None = None
    character_count: int = Field(ge=0)
    word_count: int = Field(ge=0)
    metadata: dict[str, object] | None = None
    created_at: datetime


class ExtractedContainerExportResponse(ExtractedContainerResponse):
    """Full container content serialized only by the export path."""

    raw_text: str
    normalised_text: str


class ExtractedBlockResponse(ApiSchema):
    """One native or OCR viewer block with retained source provenance."""

    id: UUID
    extraction_run_id: UUID
    container_id: UUID
    parent_block_id: UUID | None = None
    block_type: ExtractedBlockType
    block_order: int = Field(ge=0)
    source_reference: str
    text: str
    normalised_text: str
    style_name: str | None = None
    heading_level: int | None = Field(default=None, ge=1, le=9)
    location: dict[str, object] | None = None
    metadata: dict[str, object] | None = None
    character_count: int = Field(ge=0)
    word_count: int = Field(ge=0)
    created_at: datetime
    content_source: ExtractedContentSource = "NATIVE"
    language_code: LanguageCode | None = None
    language_confidence: float | None = Field(default=None, ge=0, le=1)
    ocr_confidence: float | None = Field(default=None, ge=0, le=1)
    provenance: dict[str, object] = Field(default_factory=dict)


class ExtractedTableCellResponse(ApiSchema):
    id: UUID
    extracted_table_id: UUID
    row_index: int = Field(ge=0)
    column_index: int = Field(ge=0)
    row_span: int = Field(ge=1)
    column_span: int = Field(ge=1)
    coordinate: str | None = None
    text: str
    normalised_text: str
    metadata: dict[str, object] | None = None
    created_at: datetime


class ExtractedTableResponse(ApiSchema):
    """Bounded table metadata and cells for paginated API responses."""

    id: UUID
    extraction_run_id: UUID
    container_id: UUID
    source_reference: str
    table_index: int = Field(ge=0)
    row_count: int = Field(ge=0)
    column_count: int = Field(ge=0)
    metadata: dict[str, object] | None = None
    cells: list[ExtractedTableCellResponse] = Field(default_factory=list)
    created_at: datetime


class ExtractedTableExportResponse(ExtractedTableResponse):
    """Full table text serialized only by the export path."""

    raw_text: str


class ExtractedContentSearchItem(ApiSchema):
    block_id: UUID
    block_order: int = Field(ge=0)
    container_id: UUID
    container_index: int = Field(ge=0)
    container_name: str | None = None
    source_reference: str
    block_type: ExtractedBlockType
    snippet: str
    location: dict[str, object] | None = None


class ExtractedContentSearchResponse(ApiSchema):
    query: str
    total_matches: int = Field(ge=0)
    items: list[ExtractedContentSearchItem]


class ExtractedContainerListResponse(
    PaginationData[ExtractedContainerResponse]
):
    pass


class ExtractedBlockListResponse(PaginationData[ExtractedBlockResponse]):
    pass


class ExtractedTableListResponse(PaginationData[ExtractedTableResponse]):
    pass


ExtractionExportFormat = Literal["json", "txt"]

File: frontend/src/types/extractedContent.ts
import type {
  ExtractionDocumentSummary,
  ExtractionFileSummary,
  ExtractionResultSummary,
  ExtractionRevisionSummary,
  ExtractionRunStatus,
  ExtractorType,
} from './extraction';
import type { PaginatedData } from './masterData';
import type { LanguageCode } from './languageDetection';

export type ExtractedContainerType =
  'PDF_PAGE' | 'DOCX_BODY' | 'DOCX_HEADER' | 'DOCX_FOOTER' | 'XLSX_WORKSHEET';

export type ExtractedBlockType =
  | 'TEXT'
  | 'PARAGRAPH'
  | 'HEADING'
  | 'TABLE'
  | 'TABLE_ROW'
  | 'TABLE_CELL'
  | 'HEADER'
  | 'FOOTER'
  | 'WORKSHEET_TITLE'
  | 'CELL'
  | 'MERGED_CELL'
  | 'FORMULA'
  | 'PAGE_NUMBER'
  | 'UNKNOWN';

export interface ExtractionRun extends ExtractionResultSummary {
  extractionJobId: string;
  document: ExtractionDocumentSummary;
  revision: ExtractionRevisionSummary;
  file: ExtractionFileSummary;
  extractorVersion: string;
  sourceSha256Hash: string;
  sourceFileSize: number;
  contentHash: string | null;
  metadata: Record<string, unknown> | null;
  startedAt: string;
  completedAt: string;
  createdAt: string;
  requestedBy?: {
    id: string;
    name: string;
  } | null;
  reExtractionReason?: string | null;
  isLatest: boolean;
}

export interface ExtractionRunHistoryItem {
  id: string;
  extractionJobId: string;
  extractorType: ExtractorType;
  extractorVersion: string;
  status: ExtractionRunStatus;
  sourceSha256Hash: string;
  contentHash: string | null;
  summary: ExtractionResultSummary;
  requestedBy: {
    id: string;
    name: string;
  } | null;
  reExtractionReason: string | null;
  warnings: string[];
  completedAt: string;
  isLatest: boolean;
}

export interface ExtractedContainer {
  id: string;
  extractionRunId: string;
  containerType: ExtractedContainerType;
  containerIndex: number;
  name: string | null;
  title: string | null;
  characterCount: number;
  wordCount: number;
  metadata: Record<string, unknown> | null;
  createdAt: string;
}

export interface ExtractedBlock {
  id: string;
  extractionRunId: string;
  containerId: string;
  parentBlockId: string | null;
  blockType: ExtractedBlockType;
  blockOrder: number;
  sourceReference: string;
  text: string;
  normalisedText: string;
  styleName: string | null;
  headingLevel: number | null;
  location: Record<string, unknown> | null;
  metadata: Record<string, unknown> | null;
  characterCount: number;
  wordCount: number;
  contentSource?: 'NATIVE' | 'OCR';
  languageCode?: LanguageCode | null;
  languageConfidence?: number | null;
  ocrConfidence?: number | null;
  createdAt: string;
}

export interface ExtractedTableCell {
  id: string;
  extractedTableId: string;
  rowIndex: number;
  columnIndex: number;
  rowSpan: number;
  columnSpan: number;
  coordinate: string | null;
  text: string;
  normalisedText: string;
  metadata: Record<string, unknown> | null;
  createdAt: string;
}

export interface ExtractedTable {
  id: string;
  extractionRunId: string;
  containerId: string;
  sourceReference: string;
  tableIndex: number;
  rowCount: number;
  columnCount: number;
  metadata: Record<string, unknown> | null;
  cells: ExtractedTableCell[];
  createdAt: string;
}

export interface ExtractionSearchResult {
  blockId: string;
  blockOrder: number;
  containerId: string;
  containerIndex: number;
  containerName: string | null;
  sourceReference: string;
  blockType: ExtractedBlockType;
  snippet: string;
  location: Record<string, unknown> | null;
}

export interface ExtractedContentSearchResponse {
  query: string;
  totalMatches: number;
  items: ExtractionSearchResult[];
}

export interface ExtractionContainerListParams {
  containerType?: ExtractedContainerType;
  search?: string;
  page: number;
  pageSize: number;
}

export interface ExtractionBlockListParams {
  containerId?: string;
  blockType?: ExtractedBlockType;
  contentSource?: 'NATIVE' | 'OCR';
  languageCode?: LanguageCode;
  search?: string;
  page: number;
  pageSize: number;
  sortOrder?: 'asc' | 'desc';
}

export interface ExtractionTableListParams {
  containerId?: string;
  search?: string;
  includeCells?: boolean;
  page: number;
  pageSize: number;
}

export interface ExtractionSearchParams {
  q: string;
  page?: number;
  pageSize?: number;
}

export type ExtractedContainerList = PaginatedData<ExtractedContainer>;
export type ExtractedBlockList = PaginatedData<ExtractedBlock>;
export type ExtractedTableList = PaginatedData<ExtractedTable>;
export type ExtractionRunHistory = PaginatedData<ExtractionRunHistoryItem>;

export interface ExtractionDownload {
  blob: Blob;
  fileName: string | null;
}

export interface ExtractionFileState {
  runId: string;
  status: ExtractionRunStatus;
  extractorType: ExtractorType;
  completedAt: string;
  totalBlocks: number;
  totalCharacters: number;
  requiresOcr: boolean;
}
